### PR TITLE
Add various Listener changes and improvements

### PR DESCRIPTION
**Pull request by Caelin**

### CODE REVIEW

**Review by ghost — APPROVED** [review]

**Comment by gravityfox** [comment]
So my only concern with the attack entity listener is that you're adding code to fix a problem that didn't use to exist. The code also adds a measurable amount of extra overhead and even has some ambiguity.

I'll take a look at this in a few and merge it in with some changes, but I believe the issue with the arrows lies in how I compute the blame for the event, which is something I knew that I'd have to address with API 7, but haven't gotten around to doing yet. And if that doesn't fix the issue, I'll let the sponge team know.

Basically, item frames MUST throw a damage event before they're destroyed by things other than player clicks. They are entities and their destruction must be cancellable, and the only point you can cancel that is the DamageEntityEvent.

I think I might still end up adding support for an "attack" flag that is fired first, cause in many cases we don't actually even care to compute the damage event at all, but we'll see.

**Comment by Caelin** [comment]
Fox I agree that a DamageEntityEvent should be fired for the item frames but they aren't currently and my guess would be it isn't happening because item frames really don't have HP so why do the DamageEntityEvent for them but this is only a guess.  This could still be a bug with Sponge or something that is breaking because of Pixelmon or some other mod I am using.  Most of these changes are for my specific server to make things work on it so with vanilla Sponge some of these changes may not be needed so you will have to test that I guess.

Also I agree adding an attack flag is a good idea because yes most of the time when you are blocking the damage you don't care about the amount of damage you just want to block it so allowing that to process is a waste of resources.

**Comment by gravityfox** [comment]
So I may or may not have actually found the cause of arrows getting past protections, that had to do with a fundamental flaw in the flags system.

I am now fixing that issue, but I will also still include `attack` as a flag, for safety's sake.
